### PR TITLE
Manage calendars as set of Dates

### PR DESCRIPTION
**Pull request by datanel**

### CODE REVIEW

File: src/collection.rs [hunk 1]
```diff
@@ -311,6 +311,12 @@ impl<T: Id<T>> CollectionWithId<T> {
         }
     }
 
+    /// Returns a reference to the object corresponding to the
+    /// identifier.
```
Comment (datanel) [inline src/collection.rs]: this comment is wrong.
@TeXitoi can you help me to find a description and a good example please?

File: src/common_format.rs [hunk 2]
```diff
@@ -31,34 +34,133 @@ pub struct CalendarDate {
...
+                    }
+                    ExceptionType::Remove => {
+                        calendar.dates.remove(&calendar_date.date);
+                    }
+                });
+            is_inserted.unwrap_or_else(|| {
+                if calendar_date.exception_type == ExceptionType::Add {
+                    let mut dates = BTreeSet::new();
+                    dates.insert(calendar_date.date);
+                    calendars.push(objects::Calendar {
```
Comment (datanel) [inline src/common_format.rs]: here I have a warning
```
warning: unused `std::result::Result` which must be used
   --> src/common_format.rs:128:21
    |
128 | /                     calendars.push(objects::Calendar {
129 | |                         id: calendar_date.service_id,
130 | |                         dates,
131 | |                     });
    | |_______________________^
    |
    = note: #[warn(unused_must_use)] on by default
```
I don't know how to remove it

Comment (TeXitoi) [reply src/common_format.rs]: adding a `?` at the end.

File: src/collection.rs (hunk 1 above)
Comment (TeXitoi) [reply src/collection.rs]: I think you should take inspiration if index_mut. Thus, something like 

> Access to a mutable reference of the object corresponding to the identifier.
>
> The `drop` of the proxy object panic if the identifier is modified to an identifier already on the collection.

and an example close to index_mut

File: src/objects.rs [hunk 3]
```diff
@@ -775,32 +775,14 @@ pub enum ExceptionType {
     Remove,
 }
 
-pub type CalendarDates = Vec<(Date, ExceptionType)>;
+use std::collections::BTreeSet;
 
 #[derive(Serialize, Deserialize, Debug, PartialEq)]
 pub struct Calendar {
-    #[serde(rename = "service_id")]
+    #[serde(skip)]
```
Comment (TeXitoi) [inline src/objects.rs]: please remove these skip. It's useless for gtfs/nffs, and may be useful someday if we want to support serde on the ntm::Model.

File: src/common_format.rs (hunk 2 above)
Comment (datanel) [reply src/common_format.rs]:
```
--> src/common_format.rs:128:21
    |
128 | /                     calendars.push(objects::Calendar {
129 | |                         id: calendar_date.service_id,
130 | |                         dates,
131 | |                     })?;
    | |_______________________^ cannot use the `?` operator in a function that returns `(
```

Comment (datanel) [reply src/common_format.rs]: works with unwrap as suggested by @prhod

File: src/common_format.rs [hunk 4]
```diff
@@ -31,34 +34,135 @@ pub struct CalendarDate {
...
+            valid_days.push(chrono::Weekday::Sun);
+        }
+
+        valid_days
+    }
+
+    fn get_valid_dates(&self) -> BTreeSet<Date> {
+        let valid_days = self.get_valid_days();
+        let duration = self.end_date - self.start_date;
+        (0..duration.num_days() + 1)
```
[the remark this answers is not in the document]
Comment (datanel) [reply src/common_format.rs]: yes because the upper bound of a range in rust is exclusive

**Pull request merged**